### ISSUE TITLE
 space in the prompt disappears after enter echo

### ISSUE THREAD

**Issue by simaek** [chrisant996/clink#69]
@chrisant996 
After you enter the keyword 'echo', enter any character, such as 'a', the leading whitespace will disappear. Maybe there are other keywords can cause this issue that I have not found out yet.

Cmder: v1.3.17
Clink: v1.1.26

**Comment by chrisant996** [comment]
Thank you for reporting this!

I don't understand what's being reported; can you add information and elaborate?
Here is a gif of me trying to reproduce the issue:
![echo_a](https://user-images.githubusercontent.com/17440311/107269324-4d095900-69fe-11eb-8549-08d8c6804687.gif)

I don't see leading whitespace disappear when I try it, so there must be something different in your installation.
- It doesn't repro with `clink set clink.colorize_input on` or `off`.
- It doesn't repro with `echo a` or `echoa`.

Are you able to reproduce the issue in a fresh install of Cmder 1.3.17 + Clink 1.1.26?  (That's what I tried.)  Are there some additional configurations that cause the problem to start occurring?

**Comment by chrisant996** [comment]
Does it only repro if the prompt contains UTF8 characters?
Can you share a screen shot and indicate where leading spaces have disappeared?

Thanks!  I want to track this down as soon as possible.

**Comment by simaek** [comment]
@chrisant996 
![TestLiveTemplates](https://user-images.githubusercontent.com/35521828/107312726-07b35d00-6acc-11eb-8d47-f55fbbfa2b95.gif)

**Comment by chrisant996** [comment]
It seems we are both using completely fresh installs, but the behavior is different.
So clearly there must be something else affecting it -- but I don't know how to guess what.

Would you be comfortable following steps [here](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/time-travel-debugging-record) to attach to the cmd.exe process and capture a Time Travel Trace of the problem?  The steps require WinDbg Preview ([install from here](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/windbg-install-preview)).

Basically you'd launch Cmder, attach WinDbg Preview for Time Travel Tracing, then type "echod" and reproduce the problem, then Stop the tracing.  It would write out a cmd.run file (or cmd01.run or whatever).  If we can get the .run file to me, then I could debug the .run file and observe exactly what happened on your computer.  That would make it pretty easy to figure out and fix the problem.  But, obviously, there would be some extra effort on your side, plus you'd need to be comfortable sending me the .run file.

**Comment by simaek** [comment]
Here is the TTD .run file:

https://drive.google.com/file/d/1USJcjL41Hh-YFLCIW2LJ44wsMZ0__L5l/view?usp=sharing

**Comment by chrisant996** [comment]
Thank you -- that's the right procedure, but it captured a trace for ConEmu64.exe, and I need a trace of cmd.exe (Clink runs as part of cmd.exe).

Would you mind doing the same procedure again, but attaching to cmd.exe instead?  The procedure was correct, it just attached to the wrong process -- the .run file is working at my end, but it's for a different process so it doesn't provide diagnostic information about the Clink problem.

I appreciate your help and your time very much; thank you for helping to track down what's going on here!

**Comment by simaek** [comment]
I’m happy to help.

https://drive.google.com/file/d/1D4BFGcmgsU8AByptkwd71l5-H0SUrQJw/view?usp=sharing

**Comment by chrisant996** [comment]
Perfect, that's exactly what I needed, thank you.  It will take some time to analyze it, but I should be able to track down the cause from that trace.

**Comment by chrisant996** [comment]
Ok this is very strange.  In the trace I can observe it outputting everything correctly, and when I step through the trace and a live debugging session at the same time, they seem to behave exactly the same.  Obviously something is happening that I'm missing or can't inspect.

The symptoms are consistent with mishandled multi-byte UTF8 text.  But the only UTF8 text I can see is the lambda character, and the code seems to be handling it correctly, and handling it the same in the trace as in a live debugging session (where the output works as expected, per the GIF I shared).

**So, may I please request some additional information?**

Please consider doing as many of the following as you're willing to:

1. Can you send me a couple files?  Please run `clink info` and then:
  - Look for the "settings" line and please send the corresponding settings file.  This will make sure that our computers have the same fresh settings file.  For example, maybe the settings file gets generated differently depending on the current OS locale.
  - Look for the "inputrc" section of lines, and please send all of the files that say "(exists)" next to them.  The inputrc file by default is shared across all installs (and even potentially with Bash and other things), so a pre-existing inputrc file can potentially unexpectedly affect fresh installs.  Maybe there are differences between our inputrc files.
2. Can you try overriding the terminal emulation mode?  Please try:
  - `clink set terminal.emulation emulate` and then try the `echoa` thing again.  Does that work?
  - `clink set terminal.emulation native` and then try the `echoa` thing again.  Does that work?
  - `clink set terminal.emulation clear` will reset it to the default (which is "auto").
3. I can see that the font in your Cmder is different than the font in my Cmder.  Can you find what font face and size Cmder is configured to use?
4. Can you run `set > envvars.txt` and send the envvars.txt file, after confirming you're comfortable sharing its contents?
5. Can you try using Clink 1.1.**25** and see what happens?  This is _before_ it fixed UTF8 input key bindings.  That shouldn't affect output at all, but it could be informative to check.
6. Can you try using Clink 1.1.**17** and see what happens?  This is _before_ it picked up Readline 8.1, which introduced some big changes to the output routines.

If none of those help, then I'll have to try sending a .lua script that can capture some more information such as the exact prompt text.  Or maybe even build a diagnostic feature that can capture all of the terminal output, _combined with OS console APIs to confirm the cursor position at each point_, to double check that the console host is handling the ANSI escape codes as expected.

**Also, here is what I observe in the trace:**

First, I want to assure you that I believe you're experiencing the reported problem.  I have no doubt of that.  I'm just having trouble discovering where and why it's happening (and whether it might be going wrong outside of Clink's control).

- The trace begins.
- Input `e` `c` `h` `o`, and each redisplays the input line correctly.
- Input `a` and it redisplay the input line correctly:
  - It decides that `a` got added, and that the color of `echo` changed.
  - It decides that `echoa` needs to be printed, and chooses the appropriate color.
  - It emits `\x0d` (CR) to move cursor to first column.
  - It emits `\x1b[C` (ESC[C) to move cursor right one position (skipping past the prompt's lambda character).
  - It again emits `\x1b[C` (ESC[C) to move cursor right one more position (skipping past the prompt's space character).
    - That's the very strange part:  it's explicitly skipping past the space, and yet what you observe on the screen (and the GIF you sent) indicates it somehow isn't successfully skipping past the space.
  - At this point, the cursor should be at the second position, where the `echoa` should start.  But there's no way to tell where ConEmu may have actually put the cursor.
  - It emits the correct string for the correctly colored `echoa`.
  - Then it's done.
- Input a bunch of backspaces, which all redisplay the input line correctly.
- Input `e` `c` `h` `o`, each of which redisplays the input line correctly.
- Input one backspace, which redisplays the input line correctly.
- The trace ends.

**Comment by chrisant996** [comment]
@simaek please read the preceding reply; it has a lot of info, and some requests.

In addition, I published a new [v1.1.27](https://github.com/chrisant996/clink/releases/tag/v1.1.27) build which has a new setting to log all terminal input and output to the `clink.log` file.

If you don't mind, please considering using v1.1.27 and try the following:
- Run `clink set log.rl_terminal 1` to turn on terminal logging.
- Do the `echoa` thing and reproduce the problem.
- Run `clink set log.rl_terminal 0` to turn off terminal logging.
- Use `clink info` to find the log file, and send the log file.

That will let us see not only the input and output from Clink (which I can actually see in the TTD trace, with some effort), but also the actual cursor position (which is not available in the TTD trace).  That will enable cross-referencing everything to confirm where something is going wrong, as well as a better sense of specifically what is going wrong.

Thank you your continued support!

**Comment by simaek** [comment]
> Ok this is very strange. In the trace I can observe it outputting everything correctly, and when I step through the trace and a live debugging session at the same time, they seem to behave exactly the same. Obviously something is happening that I'm missing or can't inspect.
> 
> The symptoms are consistent with mishandled multi-byte UTF8 text. But the only UTF8 text I can see is the lambda character, and the code seems to be handling it correctly, and handling it the same in the trace as in a live debugging session (where the output works as expected, per the GIF I shared).
> 
> **So, may I please request some additional information?**
> 
> Please consider doing as many of the following as you're willing to:
> 
> 1. Can you send me a couple files?  Please run `clink info` and then:
> 
> * Look for the "settings" line and please send the corresponding settings file.  This will make sure that our computers have the same fresh settings file.  For example, maybe the settings file gets generated differently depending on the current OS locale.
> * Look for the "inputrc" section of lines, and please send all of the files that say "(exists)" next to them.  The inputrc file by default is shared across all installs (and even potentially with Bash and other things), so a pre-existing inputrc file can potentially unexpectedly affect fresh installs.  Maybe there are differences between our inputrc files.
> 
> 1. Can you try overriding the terminal emulation mode?  Please try:
> 
> * `clink set terminal.emulation emulate` and then try the `echoa` thing again.  Does that work?
> * `clink set terminal.emulation native` and then try the `echoa` thing again.  Does that work?
> * `clink set terminal.emulation clear` will reset it to the default (which is "auto").
> 
> 1. I can see that the font in your Cmder is different than the font in my Cmder.  Can you find what font face and size Cmder is configured to use?  I've pasted a screen shot further below, to show my Fonts settings page in Cmder.
> 2. Can you run `set > envvars.txt` and send the envvars.txt file, after confirming you're comfortable sharing its contents?
> 3. Can you try using Clink 1.1.**25** and see what happens?  This is _before_ it fixed UTF8 input key bindings.  That shouldn't affect output at all, but it could be informative to check.
> 4. Can you try using Clink 1.1.**17** and see what happens?  This is _before_ it picked up Readline 8.1, which introduced some big changes to the output routines.
> 
> If none of those help, then I'll have to try sending a .lua script that can capture some more information such as the exact prompt text. Or maybe even build a diagnostic feature that can capture all of the terminal output, _combined with OS console APIs to confirm the cursor position at each point_, to double check that the console host is handling the ANSI escape codes as expected.
> 
> **Also, here is what I observe in the trace:**
> 
> First, I want to assure you that I believe you're experiencing the reported problem. I have no doubt of that. I'm just having trouble discovering where and why it's happening (and whether it might be going wrong outside of Clink's control).
> 
> * The trace begins.
> * Input `e` `c` `h` `o`, and each redisplays the input line correctly.
> * Input `a` and it redisplay the input line correctly:
>   
>   * It decides that `a` got added, and that the color of `echo` changed.
>   * It decides that `echoa` needs to be printed, and chooses the appropriate color.
>   * It emits `\x0d` (CR) to move cursor to first column.
>   * It emits `\x1b[C` (ESC[C) to move cursor right one position (skipping past the prompt's lambda character).
>   * It again emits `\x1b[C` (ESC[C) to move cursor right one more position (skipping past the prompt's space character).
>     
>     * That's the very strange part:  it's explicitly skipping past the space, and yet what you observe on the screen (and the GIF you sent) indicates it somehow isn't successfully skipping past the space.
>   * At this point, the cursor should be at the second position, where the `echoa` should start.  But there's no way to tell where ConEmu may have actually put the cursor.
>   * It emits the correct string for the correctly colored `echoa`.
>   * Then it's done.
> * Input a bunch of backspaces, which all redisplay the input line correctly.
> * Input `e` `c` `h` `o`, each of which redisplays the input line correctly.
> * Input one backspace, which redisplays the input line correctly.
> * The trace ends.
> 
> ![image](https://user-images.githubusercontent.com/17440311/107595215-dbc9d180-6bc8-11eb-8832-564b512449f5.png)

No problem. I want to solve this problem just like you.

1. My Clink settings is like this:

   ```
   # name: Sets how command history expansion is applied
   # type: enum
   # options: off,on,not_squoted,not_dquoted,not_quoted
   history.expand_mode = not_dquoted
   
   # name: Skip adding lines prefixed with whitespace
   # type: boolean
   history.ignore_space = False
   
   # name: The number of history lines to save
   # type: integer
   history.max_lines = 10000
   
   # name: Share history between instances
   # type: boolean
   history.shared = True
   ```

   I checked, there is no pre-existing inputrc file on my computer.

2. I tried overriding the terminal emulation mode, no matter it is set to 'emulate' or 'native', the problem will be reproduced.

![TestLiveTemplates](https://user-images.githubusercontent.com/35521828/107606208-5d286f00-6c70-11eb-9be1-39fb36923411.gif)


3. Here is my 'Fonts' settings page in Cmder. I used the latest version(210206) of ConEmu, but this is not the key. The default version(191012) integrated in Cmder will reproduce the problem too.

![image-20210211125249801](https://user-images.githubusercontent.com/35521828/107606045-da071900-6c6f-11eb-8038-5cad88e4edca.png)


4. The envvars.txt you need:

5. I tried using Clink 1.1.**25** and the problem was reproduced. I tried using Clink 1.1.**17** and the prompt became the original, like below. Now I use the Clink integrated in Cmder, version 1.1.10, and no such problem.


![image-20210211132338798](https://user-images.githubusercontent.com/35521828/107606039-d378a180-6c6f-11eb-9ecd-2771794ad30e.png)


Obviously, the trace shows that everything seems to be normal. But what actually happened on my computer was not.

There is a similar problem, the history record function sometimes will cause this problem too. When you use the up arrow key to call up the history record.

**Comment by simaek** [comment]
> @simaek please read the preceding reply; it has a lot of info, and some requests.
> 
> In addition, I published a new [v1.1.27](https://github.com/chrisant996/clink/releases/tag/v1.1.27) build which has a new setting to log all terminal input and output to the `clink.log` file.
> 
> If you don't mind, please considering using v1.1.27 and try the following:
> 
> * Run `clink set log.rl_terminal 1` to turn on terminal logging.
> * Do the `echoa` thing and reproduce the problem.
> * Run `clink set log.rl_terminal 0` to turn off terminal logging.
> * Use `clink info` to find the log file, and send the log file.
> 
> That will let us see not only the input and output from Clink (which I can actually see in the TTD trace, with some effort), but also the actual cursor position (which is not available in the TTD trace). That will enable cross-referencing everything to confirm where something is going wrong, as well as a better sense of specifically what is going wrong.
> 
> Thank you your continued support!

https://drive.google.com/file/d/1OGa4n6ZuMlphanFDY4abGcNxZQGFOcKI/view?usp=sharing

**Comment by chrisant996** [comment]
> > * Run `clink set log.rl_terminal 1` to turn on terminal logging.
> > * Do the `echoa` thing and reproduce the problem.
> 
> https://drive.google.com/file/d/1OGa4n6ZuMlphanFDY4abGcNxZQGFOcKI/view?usp=sharing
>
> There is a similar problem, the history record function sometimes will cause this problem too. When you use the up arrow key to call up the history record.

Excellent!  This reveals a lot.

Are you using the Beta UTF8 system locale feature in Windows 10?

When the OS prints the prompt's lambda character and space using that font, it may look like the cursor has moved two column positions.  But the cursor has actually moved _three_ columns, even though it only looks like two.  That makes it so Readline can't know where anything is really located on the screen, and eventually output will get messed up.

The similar problem with history (up arrow) is the exact same issue.  The reason it has started happening with "echoa" is because of how complicated Readline's drawing code it.  "echoa" changes the color of the preceding "echo", which triggers a case that skips past the prompt but draws the full input line -- which the up arrow will also often do.

So any time Readline draws the prompt + the input line, everything will come out ok (until it reaches the end of the screen line, and then things will go wrong).  But any time Readline does a partially optimized draw near the beginning of the line, it will skip past the prompt, and since the lambda character has an internal inconsistency (the pixel width is one cell, but the cell width is two cells) it ends up messing things up.  Any time Readline does a highly optimized draw where it only needs to print one or two characters at the end of the line, it will be fine because it doesn't need to move the cursor first.  And so on.

Here is the log from my computer when it first outputs the prompt.
It prints an escape code for bold default color, a lambda, a space, and escape codes to reset to the default color.
Notice the cursor starts at column 0, and ends up at column 2.
```
3938 LOGCURSORPOS              234 CURSORPOS 0,6
3938 puts_face_func            505 PUTSFACE "[1;39;40mλ [0m[m", 20
3938 rl_module::on_input      1505 INPUT "e", 1
3938 LOGCURSORPOS              234 CURSORPOS 2,6
```

Here is the log from your computer when it first outputs the prompt.
It prints the exact same 20 byte string as mine (the UTF8 lambda character is two bytes).
Notice the cursor starts at column 0, and ends up at column **3**.
```
2a50 LOGCURSORPOS              234 CURSORPOS 0,6
2a50 puts_face_func            505 PUTSFACE "[1;39;40mλ [0m[m", 20
2a50 rl_module::on_input      1505 INPUT "e", 1
2a50 LOGCURSORPOS              234 CURSORPOS 3,6
```

**Comment by chrisant996** [comment]
This sounds like [cmder #1051](https://github.com/Maximus5/ConEmu/issues/1051).

What version of Windows are you using, i.e. what does the `ver` command report from a console window?

**Comment by simaek** [comment]
Microsoft Windows [版本 10.0.19042.804]

**Comment by simaek** [comment]
I’m not sure if the Beta UTF8 system locale feature is enabled.

**Comment by chrisant996** [comment]
> I’m not sure if the Beta UTF8 system locale feature is enabled.

It's off by default, and requires very specific steps to enable it, so it's probably off.

**Comment by chrisant996** [comment]
> Microsoft Windows [版本 10.0.19042.804]

Ok thanks. It appears to be an OS issue, but I'll keep digging a bit to see if there's something Clink can/should do to compensate.

By the way, Clink internally uses UTF16 OS APIs, so it's extra strange if UTF8 quirks are the cause of the problem.

**Comment by chrisant996** [comment]
In the meantime, @simaek you can make 1.1.27 behave the same as 1.1.10 with respect to this issue by turning off the `clink.colorize_input` setting.

**Comment by simaek** [comment]
> In the meantime, @simaek you can make 1.1.27 behave the same as 1.1.10 with respect to this issue by turning off the `clink.colorize_input` setting.

It works, thanks.

**Comment by simaek** [comment]
On a whim, I changed the default prompt symbol 'λ' of the Cmder to '$', all the problems I encountered are solved. Your suggestion that I disable that Colorize Input feature is also effective. Maybe this is not just a problem with Clink. It's really strange. Anyway, thank you very much for your answer.

**Comment by chrisant996** [comment]
Oh sorry, I should have mentioned that changing the first character to be ASCII instead of UTF8 would completely resolve the issue.

And yes, it's not just Clink that's affected.  Bash is also affected.  And so is any program that prints UTF8 output and also uses GetConsoleScreenBufferInfo()` and `dwCursorPosition.X`.

Thank you for reporting this!

**Comment by chrisant996** [comment]
@simaek can you try this in plain CMD (no ConEmu), and in Windows Terminal?

1. Edit the prompt symbol back to lambda 'λ'
2. Launch `"C:\Windows\SYSTEM32\cmd.exe" /k "%CMDER_ROOT%\vendor\init.bat"` two ways:
  - From outside ConEmu (e.g. the Windows Run dialog, <kbd>Windows</kbd>+<kbd>R</kbd>)
  - From inside Windows Terminal
... but replace `%CMDER_ROOT%` with the path to Cmder's directory.
3. Try to reproduce this issue in each window.

This will identify whether it's a problem in Windows, a problem in the ConPTY psuedo-console APIs, or a problem in ConEmu itself.

Then I can know where to file a bug report, and I might even be able to do a little debugging myself depending on where the problem does/doesn't reproduce.

**Comment by chrisant996** [comment]
Aha! I'm able to reproduce this locally after running `chcp 932`.
It does not repro with `chcp 65001`.

@simaek if you use `chcp 65001` does that interfere with anything?

I ask because I wonder if Clink might be able to work around this issue by internally changing the codepage to 65001 (UTF8) while accepting input, and restore it after editing.  For example, does `chcp 65001` interfere with IME input?

**Comment by chrisant996** [comment]
Found it.  Will post with more details later, but here is the short version:

There are two bugs:
1. ConEmu doesn't tell the underlying console what font it will use for rendering.
2. Clink/Readline assume 1 cell width for ambiguous width characters, and should instead ask the font.

The underlying console uses a font where the lambda is 2 cells wide.
ConEmu draws the text uses a different font where the lambda is only 1 cell wide.
So it looks like the lambda is 1 cell width, and thus it's mysterious why the cursor position is wrong.

I'll fix the Clink/Readline issue, and that will make the cursor position be right -- however, if the line wraps then ConEmu will wrap it earlier than it looks like it should.  That part can only be fixed in ConEmu.

**Comment by chrisant996** [comment]
Fixed!!

Ok, it's partially fixed.  The part in Clink (Readline) is fixed.  But to fully fix the overall issue there would need to be changes in ConEmu.

CJK ambiguous width characters are now calculated correctly in code pages 932, 936, 949, and 950.
Verified in standalone CMD.exe, ConEmu, Windows Terminal, and ConsoleZ.

**Comment by chrisant996** [comment]
_(Clink v1.4.12 should solve most of the cursor position and input line glitching problems in CJK code pages.)_
